feat: Add scim as a source for nsscache

git-it127 opened the pull request:
Add in scim as a source for nsscache.

Can be tested with:
update the conf file with scim config at examples/nssache-scim.conf. Then issue the command
`nsscache update -v -c examples/nssache-scim.conf`.
```bash
# display the output of the cache
ls -la /tmp/nsscache/ && echo "--- PASSWD ---" && cat /tmp/nsscache/passwd.cache && echo -e "\n--- GROUP ---" && cat /tmp/nsscache/group.cache && echo -e "\n--- SSHKEY ---" && cat /tmp/nsscache/sshkey.cache
```

git-it127 committed:
feat: Add scim as a source for nsscache

bradbeam commented on nss_cache/sources/scimsource.py:
```diff
@@ -0,0 +1,612 @@
+            configuration["retry_max"] = self.RETRY_MAX
+        if "default_shell" not in configuration:
+            configuration["default_shell"] = self.DEFAULT_SHELL
+
```
Doesnt look like we support shadow; do we need to stub a method or anything for that?

git-it127 replied:
So SCIM doesn't have passwords or hashed passwords in the responses. Which is what the shadow map is made of. So there really isn't a point in making one since it would be blank any ways.

We should be able to get by without it since users will use their ssh keys to auth.

bradbeam committed:
feat: Add environment lookup for secrets in ldap and scim sources

This exposes two new environment variables for passing through auth
tokens in ldap and scim sources.

`NSSCACHE_LDAP_BIND_PASSWORD` will replace `ldap_bind_password` for the
`ldap` source.

`NSSCACHE_SCIM_AUTH_TOKEN` will replace `scim_auth_token` for the `scim`
source.

Signed-off-by: Brad Beam <bbeam@coreweave.com>

bradbeam committed:
feat: Add `ldap_default_shell`

This adds in a new parameter for ldap to set a default
shell if not specified. This should allow a loginShell
attribute defined in ldap to override this value.

Signed-off-by: Brad Beam <bbeam@coreweave.com>

bradbeam committed:
chore: Correct typo in comments/docs

Signed-off-by: Brad Beam <bbeam@coreweave.com>

git-it127 committed:
refactor: Add in pagination support for scim

Co-authored-by: gaurab15 <gacharya@coreweave.com>

git-it127 committed:
fix: Fix unit tests for scim source

kulkarnisamr commented on nss_cache/sources/scimsource.py:
```diff
@@ -0,0 +1,641 @@
+    name = "scim"
+
+    def __init__(self, conf, conn=None):
+        """Initialise the SCIM Data Source.
```
nit: American spellings pls 😆

kulkarnisamr replied:
this is from upstream
